### ISSUE TITLE
[Bug]: [rolling upgrade] Version 2.3 opens mmap and scrolls up to version 2.4, search is interrupted for more than 20 minutes after upgrade

### ISSUE THREAD

**Issue by elstic** [milvus-io/milvus#31745]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3-20240329-055dd7ea-amd64 ->  2.4-20240329-a8404f77e-amd64
- Deployment mode(standalone or cluster):cluster
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

After deploying 2.3 milvus with mmap turned on, the concurrent search and query continued, and the milvus recovery time was close to 30 minutes after rolling up to version 2.4.

Inserting 3x more data guarantees that mmap was used:
Data volume: 25m 128d data
Querynode Resources: 4c8g

Version 2.3 enables mmap with the following configuration:
```
queryNode:
   mmapDirPath: /var/lib/milvus/mmap
```

Configuration performed during rolling upgrades:
```
queryNode:
  mmap: 
    mmapEnabled: true
```

before upgrade:
```
fouramf-ddhmt-wzhh9-op-17-9268-milvus-datacoord-8d6d64bb-94x4p      Running     0            13m      10.104.5.44       4am-node12     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-datanode-7f96547457-9j4rh     Running     0            13m      10.104.1.155      4am-node10     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-indexcoord-7cb4cb8f4gpkqz     Running     0            13m      10.104.13.138     4am-node16     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-indexnode-644889c5f-nfkdr     Running     0            13m      10.104.5.43       4am-node12     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-proxy-bfd7d8cc-ws8c5          Running     0            13m      10.104.13.139     4am-node16     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-querycoord-5c457cff5npl7n     Running     0            13m      10.104.1.154      4am-node10     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-querynode-0-545c5d985k9h7     Running     0            13m      10.104.15.160     4am-node20     
fouramf-ddhmt-wzhh9-op-17-9268-milvus-rootcoord-7cfbdc5d5d27sbr     Running     0            13m      10.104.1.153      4am-node10     
fouramf-ddhmt-wzhh9-op-17-9268-etcd-0                               Running     0            2h3m     10.104.15.31      4am-node20     
fouramf-ddhmt-wzhh9-op-17-9268-etcd-1                               Running     0            2h3m     10.104.32.242     4am-node39     
fouramf-ddhmt-wzhh9-op-17-9268-etcd-2                               Running     0            2h3m     10.104.19.152     4am-node28     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-0                      Running     0            2h2m     10.104.15.33      4am-node20     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-1                      Running     0            2h2m     10.104.32.245     4am-node39     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-2                      Running     0            2h2m     10.104.21.104     4am-node24     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-broker-0                      Running     0            2h3m     10.104.13.65      4am-node16     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-proxy-0                       Running     0            2h3m     10.104.6.82       4am-node13     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-recovery-0                    Running     0            2h3m     10.104.13.63      4am-node16     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-0                   Running     0            2h3m     10.104.15.32      4am-node20     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-1                   Running     0            2h2m     10.104.25.206     4am-node30     
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-2                   Running     0            2h1m     10.104.19.157     4am-node28     
fouramf-ddhmt-wzhh9-op-17-9268-minio-0                              Running     0            2h3m     10.104.15.30      4am-node20     
fouramf-ddhmt-wzhh9-op-17-9268-minio-1                              Running     0            2h3m     10.104.32.240     4am-node39     
fouramf-ddhmt-wzhh9-op-17-9268-minio-2                              Running     0            2h3m     10.104.20.251     4am-node22     
fouramf-ddhmt-wzhh9-op-17-9268-minio-3                              Running     0            2h3m     10.104.21.101     4am-node24    
```

after upgrade:
```
fouramf-ddhmt-wzhh9-op-17-9268-etcd-0                             1/1     Running                           0               165m    10.104.15.31    4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-etcd-1                             1/1     Running                           0               165m    10.104.32.242   4am-node39   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-etcd-2                             1/1     Running                           0               165m    10.104.19.152   4am-node28   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-datacoord-5fd74ffb44mmvnw   1/1     Running                           0               36m     10.104.4.42     4am-node11   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-datanode-785d8656dc-2zcxb   1/1     Running                           0               2m49s   10.104.6.228    4am-node13   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-indexcoord-6ffd5c866dsgc4   1/1     Running                           0               35m     10.104.15.167   4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-indexnode-67958d7fd9w2vqh   1/1     Running                           0               38m     10.104.21.16    4am-node24   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-proxy-5d77d8cf6c-nrsmk      1/1     Running                           0               102s    10.104.27.5     4am-node31   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-querycoord-57df4f7d7zzr65   1/1     Running                           0               34m     10.104.15.168   4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-querynode-0-84df558crsr6s   1/1     Running                           0               33m     10.104.4.63     4am-node11   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-milvus-rootcoord-76d58d894dlctq7   1/1     Running                           0               37m     10.104.1.159    4am-node10   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-minio-0                            1/1     Running                           0               165m    10.104.15.30    4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-minio-1                            1/1     Running                           0               165m    10.104.32.240   4am-node39   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-minio-2                            1/1     Running                           0               165m    10.104.20.251   4am-node22   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-minio-3                            1/1     Running                           0               165m    10.104.21.101   4am-node24   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-0                    1/1     Running                           0               165m    10.104.15.33    4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-1                    1/1     Running                           0               165m    10.104.32.245   4am-node39   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-2                    1/1     Running                           0               165m    10.104.21.104   4am-node24   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-bookie-init-jgpjt           0/1     Completed                         0               165m    10.104.6.81     4am-node13   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-broker-0                    1/1     Running                           0               165m    10.104.13.65    4am-node16   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-proxy-0                     1/1     Running                           0               165m    10.104.6.82     4am-node13   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-pulsar-init-gnx56           0/1     Completed                         0               165m    10.104.13.62    4am-node16   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-recovery-0                  1/1     Running                           0               165m    10.104.13.63    4am-node16   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-0                 1/1     Running                           0               165m    10.104.15.32    4am-node20   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-1                 1/1     Running                           0               164m    10.104.25.206   4am-node30   <none>           <none>
fouramf-ddhmt-wzhh9-op-17-9268-pulsar-zookeeper-2                 1/1     Running                           0               164m    10.104.19.157   4am-node28   <none>           <none>
```

client pod : fouramf-txps7-3122742286

milvus search fails about this range:   2024-03-29 17:32:00 ~ 2024-03-29 18:02:00
![image](https://github.com/milvus-io/milvus/assets/48523564/30030dc1-b7d6-41a0-8689-56db96ed7bc7)

 upgrade config :
 
![image](https://github.com/milvus-io/milvus/assets/48523564/c1b9e2e0-44d7-4ae6-8668-d2968832af95)

### Expected Behavior

Upgrades are uninterrupted or with minimal interruption

### Steps To Reproduce

```markdown
1. deploy milvus version 2.3 with operator and enable mmap. 
2. continuously perform concurrent seach /query
3. upgrade milvus to version 2.4 on a rolling basis.
```

**Comment by elstic** [comment]
Trying to upgrade from v2.3.5 to 2.4 also interrupts for more than 20 minutes

before upgrade:
```
fouramf-snq2p-op-36-4036-milvus-datacoord-754549766c-mn55s       Running     0            51m     10.104.1.146      4am-node10     
fouramf-snq2p-op-36-4036-milvus-datanode-65bdc76bd6-6z7f6        Running     0            51m     10.104.14.242     4am-node18     
fouramf-snq2p-op-36-4036-milvus-indexcoord-8544b9746f-lf8zh      Running     0            51m     10.104.32.76      4am-node39     
fouramf-snq2p-op-36-4036-milvus-indexnode-7cc454f8cd-p7x8k       Running     0            51m     10.104.1.147      4am-node10     
fouramf-snq2p-op-36-4036-milvus-proxy-76b8f9ff76-qskgm           Running     1            51m     10.104.21.173     4am-node24     
fouramf-snq2p-op-36-4036-milvus-querycoord-c9ccb46b5-rslhp       Running     1            51m     10.104.21.172     4am-node24     
fouramf-snq2p-op-36-4036-milvus-querynode-0-59557f8c8f-hvvf4     Running     0            51m     10.104.32.77      4am-node39     
fouramf-snq2p-op-36-4036-milvus-rootcoord-758dfd77cf-vz8qb       Running     0            51m     10.104.32.75      4am-node39     
fouramf-snq2p-op-36-4036-etcd-0                                  Running     0            56m     10.104.31.14      4am-node34     
fouramf-snq2p-op-36-4036-etcd-1                                  Running     0            56m     10.104.30.188     4am-node38     
fouramf-snq2p-op-36-4036-etcd-2                                  Running     0            56m     10.104.20.166     4am-node22     
fouramf-snq2p-op-36-4036-pulsar-bookie-0                         Running     0            56m     10.104.19.207     4am-node28     
fouramf-snq2p-op-36-4036-pulsar-bookie-1                         Running     0            56m     10.104.21.154     4am-node24     
fouramf-snq2p-op-36-4036-pulsar-bookie-2                         Running     0            56m     10.104.20.170     4am-node22     
fouramf-snq2p-op-36-4036-pulsar-broker-0                         Running     0            56m     10.104.31.254     4am-node34     
fouramf-snq2p-op-36-4036-pulsar-proxy-0                          Running     0            56m     10.104.25.65      4am-node30     
fouramf-snq2p-op-36-4036-pulsar-recovery-0                       Running     0            56m     10.104.6.177      4am-node13     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-0                      Running     0            56m     10.104.30.189     4am-node38     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-1                      Running     0            55m     10.104.27.203     4am-node31     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-2                      Running     0            54m     10.104.24.61      4am-node29     
fouramf-snq2p-op-36-4036-minio-0                                 Running     0            56m     10.104.31.9       4am-node34     
fouramf-snq2p-op-36-4036-minio-1                                 Running     0            56m     10.104.19.206     4am-node28     
fouramf-snq2p-op-36-4036-minio-2                                 Running     0            56m     10.104.30.186     4am-node38     
fouramf-snq2p-op-36-4036-minio-3                                 Running     0            56m     10.104.25.66      4am-node30     
```

after upgrade:
```
NAME                                                            STATUS      RESTARTS     AGE       IP                NODE           
fouramf-snq2p-op-36-4036-milvus-datacoord-7f9c48d7bc-q2gk5      Running     0            22m       10.104.21.38      4am-node24     
fouramf-snq2p-op-36-4036-milvus-datanode-77bb4797dd-f7mz2       Running     0            22m       10.104.27.3       4am-node31     
fouramf-snq2p-op-36-4036-milvus-indexcoord-7db7b4f5c5-lrcrb     Running     0            22m       10.104.21.37      4am-node24     
fouramf-snq2p-op-36-4036-milvus-indexnode-c8f979858-rzl88       Running     0            22m       10.104.15.187     4am-node20     
fouramf-snq2p-op-36-4036-milvus-proxy-7454446576-fjzx6          Running     0            22m       10.104.27.2       4am-node31     
fouramf-snq2p-op-36-4036-milvus-querycoord-7df6f77687-qzc94     Running     0            22m       10.104.21.35      4am-node24     
fouramf-snq2p-op-36-4036-milvus-querynode-1-f6cc69654-bq77k     Running     0            22m       10.104.24.148     4am-node29     
fouramf-snq2p-op-36-4036-milvus-rootcoord-69ff5b5cdc-v22rf      Running     0            22m       10.104.15.186     4am-node20     
fouramf-snq2p-op-36-4036-etcd-0                                 Running     0            1h19m     10.104.31.14      4am-node34     
fouramf-snq2p-op-36-4036-etcd-1                                 Running     0            1h19m     10.104.30.188     4am-node38     
fouramf-snq2p-op-36-4036-etcd-2                                 Running     0            1h19m     10.104.20.166     4am-node22     
fouramf-snq2p-op-36-4036-pulsar-bookie-0                        Running     0            1h19m     10.104.19.207     4am-node28     
fouramf-snq2p-op-36-4036-pulsar-bookie-1                        Running     0            1h19m     10.104.21.154     4am-node24     
fouramf-snq2p-op-36-4036-pulsar-bookie-2                        Running     0            1h19m     10.104.20.170     4am-node22     
fouramf-snq2p-op-36-4036-pulsar-broker-0                        Running     0            1h19m     10.104.31.254     4am-node34     
fouramf-snq2p-op-36-4036-pulsar-proxy-0                         Running     0            1h19m     10.104.25.65      4am-node30     
fouramf-snq2p-op-36-4036-pulsar-recovery-0                      Running     0            1h19m     10.104.6.177      4am-node13     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-0                     Running     0            1h19m     10.104.30.189     4am-node38     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-1                     Running     0            1h18m     10.104.27.203     4am-node31     
fouramf-snq2p-op-36-4036-pulsar-zookeeper-2                     Running     0            1h17m     10.104.24.61      4am-node29     
fouramf-snq2p-op-36-4036-minio-0                                Running     0            1h19m     10.104.31.9       4am-node34     
fouramf-snq2p-op-36-4036-minio-1                                Running     0            1h19m     10.104.19.206     4am-node28     
fouramf-snq2p-op-36-4036-minio-2                                Running     0            1h19m     10.104.30.186     4am-node38     
fouramf-snq2p-op-36-4036-minio-3                                Running     0            1h19m     10.104.25.66      4am-node30     
```

client pod : fouramf-snq2p-753518866
milvus search fails about this range: 2024-03-29 17:44:00 ~ 2024-03-29 18:04:00

**Comment by elstic** [comment]
https://github.com/milvus-io/milvus/pull/31995 

This pr has a fix for upgrades, v2.3.12 Upgrade to 2.4 Interrupt for 1 minute

server:
```
issue-31964-milvus-datacoord-58648cbf89-gntb6      Running     0            21m       10.104.13.225     4am-node16     
issue-31964-milvus-datanode-9fc8dbf86-wqcnz        Running     0            21m       10.104.13.224     4am-node16     
issue-31964-milvus-indexcoord-6564f696f7-l7hkk     Running     0            21m       10.104.9.240      4am-node14     
issue-31964-milvus-indexnode-65f5f6774d-q5r6s      Running     0            21m       10.104.29.218     4am-node35     
issue-31964-milvus-proxy-55b9c8987d-8hprj          Running     0            21m       10.104.4.97       4am-node11     
issue-31964-milvus-querycoord-65bf4f6f45-4ktht     Running     0            21m       10.104.4.98       4am-node11     
issue-31964-milvus-querynode-1-7c4df8d96-l4hnt     Running     0            21m       10.104.9.241      4am-node14     
issue-31964-milvus-rootcoord-6559f55849-r4jts      Running     0            21m       10.104.29.217     4am-node35     
issue-31964-etcd-0                                 Running     0            1h43m     10.104.15.71      4am-node20     
issue-31964-etcd-1                                 Running     0            1h43m     10.104.16.89      4am-node21     
issue-31964-etcd-2                                 Running     0            1h43m     10.104.34.160     4am-node37     
issue-31964-pulsar-bookie-0                        Running     0            1h43m     10.104.15.72      4am-node20     
issue-31964-pulsar-bookie-1                        Running     0            1h43m     10.104.34.163     4am-node37     
issue-31964-pulsar-bookie-2                        Running     0            1h43m     10.104.33.83      4am-node36     
issue-31964-pulsar-broker-0                        Running     0            1h43m     10.104.16.86      4am-node21     
issue-31964-pulsar-proxy-0                         Running     0            1h43m     10.104.5.9        4am-node12     
issue-31964-pulsar-recovery-0                      Running     0            1h43m     10.104.15.65      4am-node20     
issue-31964-pulsar-zookeeper-0                     Running     0            1h43m     10.104.15.73      4am-node20     
issue-31964-pulsar-zookeeper-1                     Running     0            1h42m     10.104.23.92      4am-node27     
issue-31964-pulsar-zookeeper-2                     Running     0            1h41m     10.104.34.169     4am-node37     
issue-31964-minio-0                                Running     0            1h43m     10.104.15.70      4am-node20     
issue-31964-minio-1                                Running     0            1h43m     10.104.16.88      4am-node21     
issue-31964-minio-2                                Running     0            1h43m     10.104.34.156     4am-node37     
issue-31964-minio-3                                Running     0            1h43m     10.104.18.164     4am-node25 
```
client pod: fouramf-sr2v9-1410733272

 Interruption time within 1 minute

**Comment by elstic** [comment]
Should be this pr fix: https://github.com/milvus-io/milvus/pull/31995
v2.3.12 upgraded to 2.4, upgrade interrupted for 16s, considered reasonable  
Interruption time: 11:34:34~11:34:50

server:
```
fouramf-tnmld-op-96-3196-milvus-datacoord-59cbc855d7-5qblw       Running     0            1m      10.104.30.241     4am-node38     
fouramf-tnmld-op-96-3196-milvus-datanode-6b5b744cbb-l8kxv        Running     0            1m      10.104.4.138      4am-node11     
fouramf-tnmld-op-96-3196-milvus-indexcoord-6d8f4494b9-tmnt4      Running     0            1m      10.104.4.139      4am-node11     
fouramf-tnmld-op-96-3196-milvus-indexnode-6958d47686-5bwwz       Running     0            1m      10.104.6.62       4am-node13     
fouramf-tnmld-op-96-3196-milvus-proxy-78df799c9d-rfbxq           Running     0            1m      10.104.5.71       4am-node12     
fouramf-tnmld-op-96-3196-milvus-querycoord-8dbcdcf57-ngvvt       Running     0            1m      10.104.5.72       4am-node12     
fouramf-tnmld-op-96-3196-milvus-querynode-0-6fd4cd9549-5sl57     Running     0            1m      10.104.30.242     4am-node38     
fouramf-tnmld-op-96-3196-milvus-rootcoord-54c569b54c-8hsnz       Running     0            1m      10.104.4.137      4am-node11     
fouramf-tnmld-op-96-3196-etcd-0                                  Running     0            6m      10.104.21.165     4am-node24     
fouramf-tnmld-op-96-3196-etcd-1                                  Running     0            6m      10.104.33.168     4am-node36     
fouramf-tnmld-op-96-3196-etcd-2                                  Running     0            6m      10.104.32.45      4am-node39     
fouramf-tnmld-op-96-3196-pulsar-bookie-0                         Running     0            6m      10.104.32.46      4am-node39     
fouramf-tnmld-op-96-3196-pulsar-bookie-1                         Running     0            6m      10.104.33.174     4am-node36     
fouramf-tnmld-op-96-3196-pulsar-bookie-2                         Running     0            6m      10.104.21.169     4am-node24     
fouramf-tnmld-op-96-3196-pulsar-broker-0                         Running     0            6m      10.104.21.159     4am-node24     
fouramf-tnmld-op-96-3196-pulsar-proxy-0                          Running     0            6m      10.104.1.127      4am-node10     
fouramf-tnmld-op-96-3196-pulsar-recovery-0                       Running     0            6m      10.104.13.238     4am-node16     
fouramf-tnmld-op-96-3196-pulsar-zookeeper-0                      Running     0            6m      10.104.33.169     4am-node36     
fouramf-tnmld-op-96-3196-pulsar-zookeeper-1                      Running     0            5m      10.104.32.48      4am-node39     
fouramf-tnmld-op-96-3196-pulsar-zookeeper-2                      Running     0            4m      10.104.34.68      4am-node37     
fouramf-tnmld-op-96-3196-minio-0                                 Running     0            6m      10.104.33.170     4am-node36     
fouramf-tnmld-op-96-3196-minio-1                                 Running     0            6m      10.104.21.166     4am-node24     
fouramf-tnmld-op-96-3196-minio-2                                 Running     0            6m      10.104.32.44      4am-node39     
fouramf-tnmld-op-96-3196-minio-3                                 Running     0            6m      10.104.34.58      4am-node37  
```
